scripts/datastore_upgrade.py

```diff
@@ -0,0 +1,311 @@
+        exit(1)
+
+def start_cassandra():
+    """ Creates a monit configuration file and prompts Monit to start Cassandra. """
+    print ("Starting Cassandra...")
+    watch = "cassandra"
+    start_cmd = CASSANDRA_EXECUTABLE + " start -p " + PID_FILE
+    stop_cmd = "/usr/bin/python2 " + APPSCALE_HOME + "/scripts/stop_service.py java cassandra"
+    monit_interface.create_monit_config_file(watch, start_cmd, stop_cmd,
+        ports=[9999], match_cmd=CASSANDRA_DIR)
+    if not monit_interface.start(watch):
+        print ("Unable to start Cassandra.")
+        exit(1)
+    else:
+        print ("Successfully started Cassandra.")
+
+def start_zookeeper():
+    """ Creates a monit configuration file and prompts Monit to start ZooKeeper. """
+    print ("Starting ZooKeeper...")
+    os.system("rm -rfv /var/lib/zookeeper")
+    os.system("rm -rfv {}".format(ZK_DATA_LOCATION))
+    zk_server = "zookeeper-server"
+    if os.system("service --status-all|grep zookeeper$"):
+        zk_server = "zookeeper"
+
+    if not os.path.isdir(ZK_DATA_LOCATION):
+        print ("Initializing ZooKeeper")
+        os.system("/usr/sbin/service " + "zookeeper" + " stop")
+        os.system("mkdir -pv " + ZK_DATA_LOCATION)
+        os.system("chown -Rv zookeeper:zookeeper " + ZK_DATA_LOCATION)
+
+        if zk_server == "zookeeper-server":
+            if not os.system("/usr/sbin/service " + "zookeeper" + " init"):
+                print ("Failed to start zookeeper!")
+                exit(1)
+    os.system("ln -sfv /etc/zookeeper/conf/myid " + ZK_DATA_LOCATION + "/myid")
+    watch = "zookeeper"
+    start_cmd = "/usr/sbin/service " + "zookeeper" + " start"
+    stop_cmd = "/usr/sbin/service " + "zookeeper" + " stop"
+    match_cmd = "org.apache.zookeeper.server.quorum.QuorumPeerMain"
+    monit_interface.create_monit_config_file(watch, start_cmd, stop_cmd,
+        ports=[2181], match_cmd=match_cmd)
+    if not monit_interface.start(watch):
+        print ("Unable to start ZooKeeper")
+        exit(1)
+    else:
+        print ("Successfully started ZooKeeper")
+
+def validate_and_update_entities(datastore, ds_distributed):
+    """ Returns the SOAP server accessor to deal with application and users.
+    Args:
+        datastore: A reference to the batch datastore interface.
+    Returns:
+        A soap server accessor.
+    """
+    # Fetch all entities from the APP_ENTITY_TABLE.
+    try:
+        current_entities = datastore.range_query(APP_ENTITY_TABLE, APP_ENTITY_SCHEMA, "", "", _MAX_ENTITIES)
+    except Exception as ex:
+        print ("Exception on range query: {}".format(str(ex)))
+        exit(1)
+
+    # Remove the tombstoned entities from the list.
+    tombstoneless_entities = ds_distributed.remove_tombstoned_entities(current_entities)
+    app_ids = get_app_ids_from_entities(tombstoneless_entities)
+    entities_to_validate  = tombstoneless_entities
+
+    # Validate the entities grouped by app id.
+    for app_id in app_ids:
```

Comment:
The entities are not grouped by app ID here since you are always passing the same `entities_to_validate` var. To make this more efficient, I would change `validated_list_result` to extract the app ID. You don't want to process each entity more than once, and you definitely don't want to try looking up a transaction for a given app with an entity from another app.